# Анализ данных на Python

## Библиотека `Pandas`
#### <i>Малкова Ксения, Преподаватель ФКН НИУ ВШЭ, kemalkova@hse.ru*</i>

*Почты еще нет, но она в процессе создания :)

### Практика 

### Этап 1. Разминка

0. Считайте датафрейм по этой ссылке: 
https://raw.githubusercontent.com/rogovich/Data/master/data/titanic/train.csv

(используйте библиотеку `Pandas`)

In [1]:
# YOUR CODE HERE (ﾉ◕ヮ◕)ﾉ*:･ﾟ✧
import pandas as pd
data = pd.read_csv("https://raw.githubusercontent.com/rogovich/Data/master/data/titanic/train.csv")

In [2]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

1. Выведите 10 первых наблюдений

In [5]:
# YOUR CODE
data.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

2. Выведите 15 последних наблюдений

In [6]:
# YOUR CODE
data.tail(15)

PassengerId  Survived  Pclass  \
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                              Name     Sex   Age  SibSp  \
876                  Gustafsson, Mr. Alfred Ossian    male  20.0      0   
877                           Petroff, Mr. Nedelio    male  19.0      0   
878                             Laleff, Mr. Kristo    male   NaN      0   
879  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880   Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
881                             Markun, Mr. Johann    male  33.0      0   
882                   Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
883                  Banfield, Mr. Frederick James    male  28.0      0   
884                         Sutehall, Mr. Henry Jr    male  25.0      0   
885           Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                          Montvila, Rev. Juozas    male  27.0      0   
887                   Graham, Miss. Margaret Edith  female  19.0      0   
888       Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                          Behr, Mr. Karl Howell    male  26.0      0   
890                            Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
876      0              7534   9.8458   NaN        S  
877      0            349212   7.8958   NaN        S  
878      0            349217   7.8958   NaN        S  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
881      0            349257   7.8958   NaN        S  
882      0              7552  10.5167   NaN        S  
883      0  C.A./SOTON 34068  10.5000   NaN        S  
884      0   SOTON/OQ 392076   7.0500   NaN        S  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q

3. Выведите количество ячеек в датафрейме

In [7]:
# YOUR CODE
data.size   

10692

In [8]:
data.shape

(891, 12)

In [10]:
data.shape[0] * data.shape[1]

10692

4. Выведите названия столбцов датафрейма

In [11]:
# YOUR CODE
data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

5. Выведите общую информацию о датафрейме

In [13]:
# YOUR CODE
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Этап 2. Приступаем к работе

1. Переименуйте один столбец в данных (на ваше усмотрение)

In [16]:
# YOUR CODE HERE ✨(⌐■_■)
# раньше делали так
# но таким образом м можем переименовать только все столбцы
# data.columns = [название_1, нозвание_2, ...]

In [17]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [21]:
# data.rename(columns = {"старое значение": "новое значение"})
data = data.rename(columns = {"Survived": "survived"})
data

PassengerId  survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

2. Выведите наблюдения со 145 по 189 строку (по всем столбцам)

In [22]:
# YOUR CODE
data.iloc[145:190]

PassengerId  survived  Pclass  \
145          146         0       2   
146          147         1       3   
147          148         0       3   
148          149         0       2   
149          150         0       2   
150          151         0       2   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   
156          157         1       3   
157          158         0       3   
158          159         0       3   
159          160         0       3   
160          161         0       3   
161          162         1       2   
162          163         0       3   
163          164         0       3   
164          165         0       3   
165          166         1       3   
166          167         1       1   
167          168         0       3   
168          169         0       1   
169          170         0       3   
170          171         0       1   
171          172         0       3   
172          173         1       3   
173          174         0       3   
174          175         0       1   
175          176         0       3   
176          177         0       3   
177          178         0       1   
178          179         0       2   
179          180         0       3   
180          181         0       3   
181          182         0       2   
182          183         0       3   
183          184         1       2   
184          185         1       3   
185          186         0       1   
186          187         1       3   
187          188         1       1   
188          189         0       3   
189          190         0       3   

                                                  Name     Sex   Age  SibSp  \
145                       Nicholls, Mr. Joseph Charles    male  19.0      1   
146       Andersson, Mr. August Edvard ("Wennerstrom")    male  27.0      0   
147                   Ford, Miss. Robina Maggie "Ruby"  female   9.0      2   
148           Navratil, Mr. Michel ("Louis M Hoffman")    male  36.5      0   
149                  Byles, Rev. Thomas Roussel Davids    male  42.0      0   
150                         Bateman, Rev. Robert James    male  51.0      0   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   
156                   Gilnagh, Miss. Katherine "Katie"  female  16.0      0   
157                                    Corn, Mr. Harry    male  30.0      0   
158                                Smiljanic, Mr. Mile    male   NaN      0   
159                         Sage, Master. Thomas Henry    male   NaN      8   
160                           Cribb, Mr. John Hatfield    male  44.0      0   
161  Watt, Mrs. James (Elizabeth "Bessie" Inglis Mi...  female  40.0      0   
162                         Bengtsson, Mr. John Viktor    male  26.0      0   
163                                    Calic, Mr. Jovo    male  17.0      0   
164                       Panula, Master. Eino Viljami    male   1.0      4   
165    Goldsmith, Master. Frank John William "Frankie"    male   9.0      0   
166             Chibnall, Mrs. (Edith Martha Bowerman)  female   NaN      0   
167    Skoog, Mrs. William (Anna Bernhardina Karlsson)  female  45.0      1   
168                                Baumann, Mr. John D    male   NaN      0   
169                                      Ling, Mr. Lee    male  28.0      0   
170                          Van der hoef, Mr. Wyckoff    male  61.0      0   
171                               Rice, Master. Arthur    male   4.0      4   
172                       Johnson, Miss. Ele

3. Выведите каждое третье наблюдение со 184 по 421 строку (по всем столбцам)

In [ ]:
# YOUR CODE


4. Выведите каждое четвёртое наблюдение с 241 по 431 строку по столбцу `Age` (если вы его переименовали, используйте своё название)

In [ ]:
# YOUR CODE

### Этап 3. Фильтрация

1. Выведите только мужчин

In [24]:
# YOUR CODE HERE ╰(°▽°)╯
data[data['Sex'] == 'male']

PassengerId  survived  Pclass                            Name   Sex  \
0              1         0       3         Braund, Mr. Owen Harris  male   
4              5         0       3        Allen, Mr. William Henry  male   
5              6         0       3                Moran, Mr. James  male   
6              7         0       1         McCarthy, Mr. Timothy J  male   
7              8         0       3  Palsson, Master. Gosta Leonard  male   
..           ...       ...     ...                             ...   ...   
883          884         0       2   Banfield, Mr. Frederick James  male   
884          885         0       3          Sutehall, Mr. Henry Jr  male   
886          887         0       2           Montvila, Rev. Juozas  male   
889          890         1       1           Behr, Mr. Karl Howell  male   
890          891         0       3             Dooley, Mr. Patrick  male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
4    35.0      0      0            373450   8.0500   NaN        S  
5     NaN      0      0            330877   8.4583   NaN        Q  
6    54.0      0      0             17463  51.8625   E46        S  
7     2.0      3      1            349909  21.0750   NaN        S  
..    ...    ...    ...               ...      ...   ...      ...  
883  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
886  27.0      0      0            211536  13.0000   NaN        S  
889  26.0      0      0            111369  30.0000  C148        C  
890  32.0      0      0            370376   7.7500   NaN        Q  

[577 rows x 12 columns]

2. Выведите только мужчин в возрасте больше 25  лет

In [30]:
# YOUR CODE
data[(data['Age'] > 25) & (data['Sex'] == 'male')]

PassengerId  survived  Pclass                           Name   Sex   Age  \
4              5         0       3       Allen, Mr. William Henry  male  35.0   
6              7         0       1        McCarthy, Mr. Timothy J  male  54.0   
13            14         0       3    Andersson, Mr. Anders Johan  male  39.0   
20            21         0       2           Fynney, Mr. Joseph J  male  35.0   
21            22         1       2          Beesley, Mr. Lawrence  male  34.0   
..           ...       ...     ...                            ...   ...   ...   
881          882         0       3             Markun, Mr. Johann  male  33.0   
883          884         0       2  Banfield, Mr. Frederick James  male  28.0   
886          887         0       2          Montvila, Rev. Juozas  male  27.0   
889          890         1       1          Behr, Mr. Karl Howell  male  26.0   
890          891         0       3            Dooley, Mr. Patrick  male  32.0   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
4        0      0            373450   8.0500   NaN        S  
6        0      0             17463  51.8625   E46        S  
13       1      5            347082  31.2750   NaN        S  
20       0      0            239865  26.0000   NaN        S  
21       0      0            248698  13.0000   D56        S  
..     ...    ...               ...      ...   ...      ...  
881      0      0            349257   7.8958   NaN        S  
883      0      0  C.A./SOTON 34068  10.5000   NaN        S  
886      0      0            211536  13.0000   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN        Q  

[274 rows x 12 columns]

In [31]:
# Выведите только мужчин в возрасте от 25 до 50 лет
data[(data['Age'] > 25) & (data['Sex'] == 'male') & (data['Age'] < 50)]

PassengerId  survived  Pclass                           Name   Sex   Age  \
4              5         0       3       Allen, Mr. William Henry  male  35.0   
13            14         0       3    Andersson, Mr. Anders Johan  male  39.0   
20            21         0       2           Fynney, Mr. Joseph J  male  35.0   
21            22         1       2          Beesley, Mr. Lawrence  male  34.0   
23            24         1       1   Sloper, Mr. William Thompson  male  28.0   
..           ...       ...     ...                            ...   ...   ...   
881          882         0       3             Markun, Mr. Johann  male  33.0   
883          884         0       2  Banfield, Mr. Frederick James  male  28.0   
886          887         0       2          Montvila, Rev. Juozas  male  27.0   
889          890         1       1          Behr, Mr. Karl Howell  male  26.0   
890          891         0       3            Dooley, Mr. Patrick  male  32.0   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
4        0      0            373450   8.0500   NaN        S  
13       1      5            347082  31.2750   NaN        S  
20       0      0            239865  26.0000   NaN        S  
21       0      0            248698  13.0000   D56        S  
23       0      0            113788  35.5000    A6        S  
..     ...    ...               ...      ...   ...      ...  
881      0      0            349257   7.8958   NaN        S  
883      0      0  C.A./SOTON 34068  10.5000   NaN        S  
886      0      0            211536  13.0000   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN        Q  

[222 rows x 12 columns]

In [32]:
# Выведите только мужчин в возрасте меньше 25 или больше 50 лет
data[((data['Age'] < 25) | (data['Age'] > 50)) & (data['Sex'] == 'male')]

PassengerId  survived  Pclass                             Name   Sex  \
0              1         0       3          Braund, Mr. Owen Harris  male   
6              7         0       1          McCarthy, Mr. Timothy J  male   
7              8         0       3   Palsson, Master. Gosta Leonard  male   
12            13         0       3   Saundercock, Mr. William Henry  male   
16            17         0       3             Rice, Master. Eugene  male   
..           ...       ...     ...                              ...   ...   
861          862         0       2      Giles, Mr. Frederick Edward  male   
864          865         0       2           Gill, Mr. John William  male   
869          870         1       3  Johnson, Master. Harold Theodor  male   
876          877         0       3    Gustafsson, Mr. Alfred Ossian  male   
877          878         0       3             Petroff, Mr. Nedelio  male   

      Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
0    22.0      1      0  A/5 21171   7.2500   NaN        S  
6    54.0      0      0      17463  51.8625   E46        S  
7     2.0      3      1     349909  21.0750   NaN        S  
12   20.0      0      0  A/5. 2151   8.0500   NaN        S  
16    2.0      4      1     382652  29.1250   NaN        Q  
..    ...    ...    ...        ...      ...   ...      ...  
861  21.0      1      0      28134  11.5000   NaN        S  
864  24.0      0      0     233866  13.0000   NaN        S  
869   4.0      1      1     347742  11.1333   NaN        S  
876  20.0      0      0       7534   9.8458   NaN        S  
877  19.0      0      0     349212   7.8958   NaN        S  

[208 rows x 12 columns]

3. Выведите только мужчин в возрасте от 25 до 75 лет, которые купили билеты в первый или второй класс

In [ ]:
# YOUR CODE

4. Выведите только мужчин в возрасте от 25 до 75 лет, которые купили билеты в первый или второй класс, севшие в порту Cherbourg(С)

In [ ]:
# YOUR CODE

### Этап 4. Сортировка

1. Отсортируйте данные по возрасту

In [34]:
# YOUR CODE HERE ฅ^•ﻌ•^ฅ
data.sort_values('Age')

PassengerId  survived  Pclass                                      Name  \
803          804         1       3           Thomas, Master. Assad Alexander   
755          756         1       2                 Hamalainen, Master. Viljo   
644          645         1       3                    Baclini, Miss. Eugenie   
469          470         1       3             Baclini, Miss. Helene Barbara   
78            79         1       2             Caldwell, Master. Alden Gates   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
803    male  0.42      0      1        2625   8.5167   NaN        C  
755    male  0.67      1      1      250649  14.5000   NaN        S  
644  female  0.75      2      1        2666  19.2583   NaN        C  
469  female  0.75      2      1        2666  19.2583   NaN        C  
78     male  0.83      0      2      248738  29.0000   NaN        S  
..      ...   ...    ...    ...         ...      ...   ...      ...  
859    male   NaN      0      0        2629   7.2292   NaN        C  
863  female   NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male   NaN      0      0      345777   9.5000   NaN        S  
878    male   NaN      0      0      349217   7.8958   NaN        S  
888  female   NaN      1      2  W./C. 6607  23.4500   NaN        S  

[891 rows x 12 columns]

In [35]:
data.sort_values('Age', ascending = False)

PassengerId  survived  Pclass                                      Name  \
630          631         1       1      Barkworth, Mr. Algernon Henry Wilson   
851          852         0       3                       Svensson, Mr. Johan   
493          494         0       1                   Artagaveytia, Mr. Ramon   
96            97         0       1                 Goldschmidt, Mr. George B   
116          117         0       3                      Connors, Mr. Patrick   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
630    male  80.0      0      0       27042  30.0000   A23        S  
851    male  74.0      0      0      347060   7.7750   NaN        S  
493    male  71.0      0      0    PC 17609  49.5042   NaN        C  
96     male  71.0      0      0    PC 17754  34.6542    A5        C  
116    male  70.5      0      0      370369   7.7500   NaN        Q  
..      ...   ...    ...    ...         ...      ...   ...      ...  
859    male   NaN      0      0        2629   7.2292   NaN        C  
863  female   NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male   NaN      0      0      345777   9.5000   NaN        S  
878    male   NaN      0      0      349217   7.8958   NaN        S  
888  female   NaN      1      2  W./C. 6607  23.4500   NaN        S  

[891 rows x 12 columns]

In [36]:
data.sort_values(['Pclass', 'Age'], ascending=[True, True])

PassengerId  survived  Pclass                                      Name  \
305          306         1       1            Allison, Master. Hudson Trevor   
297          298         0       1              Allison, Miss. Helen Loraine   
445          446         1       1                 Dodge, Master. Washington   
802          803         1       1       Carter, Master. William Thornton II   
435          436         1       1                 Carter, Miss. Lucile Polk   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex    Age  SibSp  Parch      Ticket      Fare    Cabin Embarked  
305    male   0.92      1      2      113781  151.5500  C22 C26        S  
297  female   2.00      1      2      113781  151.5500  C22 C26        S  
445    male   4.00      0      2       33638   81.8583      A34        S  
802    male  11.00      1      2      113760  120.0000  B96 B98        S  
435  female  14.00      1      2      113760  120.0000  B96 B98        S  
..      ...    ...    ...    ...         ...       ...      ...      ...  
859    male    NaN      0      0        2629    7.2292      NaN        C  
863  female    NaN      8      2    CA. 2343   69.5500      NaN        S  
868    male    NaN      0      0      345777    9.5000      NaN        S  
878    male    NaN      0      0      349217    7.8958      NaN        S  
888  female    NaN      1      2  W./C. 6607   23.4500      NaN        S  

[891 rows x 12 columns]

In [38]:
data.sort_values(['Pclass', 'Age'], ascending=[False, True])

PassengerId  survived  Pclass  \
803          804         1       3   
469          470         1       3   
644          645         1       3   
164          165         0       3   
172          173         1       3   
..           ...       ...     ...   
766          767         0       1   
793          794         0       1   
815          816         0       1   
839          840         1       1   
849          850         1       1   

                                             Name     Sex   Age  SibSp  Parch  \
803               Thomas, Master. Assad Alexander    male  0.42      0      1   
469                 Baclini, Miss. Helene Barbara  female  0.75      2      1   
644                        Baclini, Miss. Eugenie  female  0.75      2      1   
164                  Panula, Master. Eino Viljami    male  1.00      4      1   
172                  Johnson, Miss. Eleanor Ileen  female  1.00      1      1   
..                                            ...     ...   ...    ...    ...   
766                     Brewe, Dr. Arthur Jackson    male   NaN      0      0   
793                      Hoyt, Mr. William Fisher    male   NaN      0      0   
815                              Fry, Mr. Richard    male   NaN      0      0   
839                          Marechal, Mr. Pierre    male   NaN      0      0   
849  Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female   NaN      1      0   

       Ticket     Fare Cabin Embarked  
803      2625   8.5167   NaN        C  
469      2666  19.2583   NaN        C  
644      2666  19.2583   NaN        C  
164   3101295  39.6875   NaN        S  
172    347742  11.1333   NaN        S  
..        ...      ...   ...      ...  
766    112379  39.6000   NaN        C  
793  PC 17600  30.6958   NaN        C  
815    112058   0.0000  B102        S  
839     11774  29.7000   C47        C  
849     17453  89.1042   C92        C  

[891 rows x 12 columns]

2. Отсортируйте данные по возрасту по убыванию

In [40]:
# YOUR CODE
data.sort_values('Age', ascending=False)

PassengerId  survived  Pclass                                      Name  \
630          631         1       1      Barkworth, Mr. Algernon Henry Wilson   
851          852         0       3                       Svensson, Mr. Johan   
493          494         0       1                   Artagaveytia, Mr. Ramon   
96            97         0       1                 Goldschmidt, Mr. George B   
116          117         0       3                      Connors, Mr. Patrick   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
630    male  80.0      0      0       27042  30.0000   A23        S  
851    male  74.0      0      0      347060   7.7750   NaN        S  
493    male  71.0      0      0    PC 17609  49.5042   NaN        C  
96     male  71.0      0      0    PC 17754  34.6542    A5        C  
116    male  70.5      0      0      370369   7.7500   NaN        Q  
..      ...   ...    ...    ...         ...      ...   ...      ...  
859    male   NaN      0      0        2629   7.2292   NaN        C  
863  female   NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male   NaN      0      0      345777   9.5000   NaN        S  
878    male   NaN      0      0      349217   7.8958   NaN        S  
888  female   NaN      1      2  W./C. 6607  23.4500   NaN        S  

[891 rows x 12 columns]

3. Отсортируйте данные сперва по возрасту (по убыванию), потом по классу пассажира (по возрастанию)

In [ ]:
# YOUR CODE

4. Удалите дубликаты по переменной `Age`

In [41]:
# YOUR CODE
data.drop_duplicates()

PassengerId  survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [42]:
data.drop_duplicates('Pclass')

PassengerId  survived  Pclass  \
0            1         0       3   
1            2         1       1   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C  
9      0     237736  30.0708   NaN        C